### PR TITLE
Stabilize prod backward graphs in low precision

### LINKED ISSUE #193012: `torch.autograd.functional.hessian` returns `inf`/`nan` in float16 where `torch.func.hessian` returns the correct value
### 🐛 Describe the bug

The double backward through `torch.prod` forms `grad_output / input` as a standalone
intermediate. When an input element is small, that quotient overflows the dtype —
even though it is immediately multiplied by a small factor, so the Hessian entry it
contributes to is an ordinary in-range number.

The result is that `torch.autograd.functional.hessian` and
`torch.autograd.functional.vhp` return `inf`/`nan` in float16, while
`torch.func.hessian`, `torch.func.jacrev(torch.func.jacrev(f))` and
`torch.func.jacfwd(torch.func.jacrev(f))` all return the correct finite value at
the same point, in the same dtype. `autograd.functional.hvp` is also unaffected. This bug is not float16-specific, it also applies to other float dtype such as float32 and float64.

## Minimal Reproducer

```python
import torch

f = lambda x: 4000.0 * torch.prod(x)
x = torch.tensor([0.03, 1.0, 1.0, 1.0], dtype=torch.float16, requires_grad=True)

print(torch.autograd.functional.hessian(f, x)[0, 1])   # inf
print(torch.func.hessian(f)(x.detach())[0, 1])         # 4000.0
# true H[0,1] = C * prod / (x0 * x1) = 4000.0
```

### Output

```
tensor(inf, dtype=torch.float16)
tensor(4000., dtype=torch.float16)
```

### Versions

Collecting environment information...
PyTorch version: 2.14.0.dev20260809+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 16.0.6 (https://github.com/llvm/llvm-project.git 7cbf1a2591520c2491aa35339f227775f4d3adf6)
CMake version: version 4.1.0
Libc version: glibc-2.39

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-134-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No devices found.
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_adv.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_cnn.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_precompiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_graph.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_heuristic.so.9.8.0
/usr/local/cuda-12.8/targets/x86_64-linux/lib/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  384
On-line CPU(s) list:                     0-383
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      2
Core(s) per socket:                      96
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU(s) scaling MHz:                      41%
CPU max MHz:                             3709.3569
CPU min MHz:                             1500.0000
BogoMIPS:                                4799.77
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc amd_ibpb_ret arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d debug_swap ibpb_exit_to_user
Virtualization:                          AMD-V
L1d cache:                               6 MiB (192 instances)
L1i cache:                               6 MiB (192 instances)
L2 cache:                                192 MiB (192 instances)
L3 cache:                                768 MiB (24 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-95,192-287
NUMA node1 CPU(s):                       96-191,288-383
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable: Clear CPU buffers attempted, no microcode
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] torch==2.14.0.dev20260809+cpu
[pip3] torchvision==0.29.0.dev20260810+cpu
[conda] numpy                     2.2.6                    pypi_0    pypi
[conda] torch                     2.14.0.dev20260809+cpu          pypi_0    pypi
[conda] torchvision               0.29.0.dev20260810+cpu          pypi_0    pypi

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by HU-qingqing**
> Fixes #193012 by routing prod backward graphs through `prod_safe_zeros_backward` when autograd records a backward graph, avoiding low-precision overflow in higher-order derivatives and functorch transforms.

> AI-assisted summary:
>
> ## Summary
>
> When autograd is recording a backward graph, route both `prod_backward`
> overloads through the division-free cumulative-product implementation.
> This avoids overflow from materializing `result / input` while preserving the fast path when a backward graph is not being recorded.
>
> Add float16 regression coverage for Hessian and VHP results and verify the
> expected result from `torch.func.hessian`.

> ## Checklist
>
> - [ ] Passes lint (`spin fixlint`) - not run
> - [x] Added/updated tests
> - [x] Updated documentation - not applicable
> - [x] Included benchmark results

> ## BC-breaking?
>
> No.

> ## Test plan
>
> - `python test/test_autograd.py -k prod_higher_order_stability -v`
> - H200 CUDA performance benchmark for `torch.func.grad(torch.prod)`

> Human review note: I reviewed the following Codex-assisted update and am including it
> because the `GradMode` guard affects first-order functorch transforms that build a
> backward graph, not only explicitly requested second derivatives.

> Codex-assisted PR description update:
>
> This change prioritizes numerical correctness for backward graphs over the
> performance of the no-zero `prod` fast path. The safe zero-aware formula is now
> selected whenever autograd is recording the backward, so the tradeoff applies to
> `torch.func.grad`, `torch.func.vjp` called with graph creation enabled,
> `torch.func.jacrev`, and `torch.autograd.grad(..., create_graph=True)`, even when
> the caller only requests a first derivative. These paths replace the
> `result / input` intermediate with the safe formulation based on concatenation,
> two cumulative products, and reverse operations. Ordinary
> `torch.autograd.grad(..., create_graph=False)` retains the fast path.
>
> Performance was measured on an NVIDIA H200 with CUDA 12.8 using the current
> implementation. The patched `torch.func.grad(torch.prod)` path checks `GradMode`
> before the zero scan and uses the safe no-division formula. The control uses
> `torch.autograd.grad(..., create_graph=False)` and includes the fast-path zero scan.
>
> | Elements | Patched CUDA | Control CUDA | Ratio |
> | ---: | ---: | ---: | ---: |
> | 1M | 0.282 ms | 0.131 ms | 2.2x |
> | 4M | 0.271 ms | 0.134 ms | 2.0x |
> | 16M | 0.494 ms | 0.232 ms | 2.1x |
>
> Each value is the median of 10 timed iterations after 3 warmup iterations, with CUDA synchronization before and after each measurement.
>
> Benchmark command:
>
> ```bash
> python - <<'PY'
> import time
> import torch
> from torch.func import grad
>
> for n in (1_000_000, 4_000_000, 16_000_000):
>     x = torch.ones(n, device="cuda")
>     f = grad(torch.prod)
>     x_control = x.detach().requires_grad_()
>
>     def patched():
>         return f(x)
>
>     def control():
>         return torch.autograd.grad(
>             torch.prod(x_control), x_control, create_graph=False
>         )[0]
>
>     for name, fn in (("patched", patched), ("control", control)):
>         for _ in range(3):
>             fn()
>             torch.cuda.synchronize()
>         times = []
>         for _ in range(10):
>             torch.cuda.synchronize()
>             start = time.perf_counter()
>             fn()
>             torch.cuda.synchronize()
>             times.append((time.perf_counter() - start) * 1000)
>         print(n, name, sorted(times)[len(times) // 2])
> ```

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section
